core/src/main/java/hudson/util/RobustReflectionConverter.java

```diff
@@ -89,7 +99,38 @@ public RobustReflectionConverter(Mapper mapper, ReflectionProvider reflectionPro
     }
 
     void addCriticalField(Class<?> clazz, String field) {
-        criticalFields.add(clazz.getName() + '#' + field);
+        // Lock the write lock
+        criticalFieldsLock.writeLock().lock();
+        try {
+            // If the class already exists, then add a new field, otherwise
+            // create the hash map field
+            if (!criticalFields.containsKey(field)) {
+                criticalFields.put(field, new HashSet<String>());
+            }
+            criticalFields.get(field).add(clazz.getName());
+        }
+        finally {
+            // Unlock
+            criticalFieldsLock.writeLock().unlock();
+        }
+    }
+    
+    private boolean hasCriticalField(Class<?> clazz, String field) {
+        // Lock the write lock
+        criticalFieldsLock.readLock().lock();
+        try {
```

Comment:
fine until something else wants to call `hasCriticalField`.  I think any further optimization like that would need to be justified - I can't see it saving that much time here? (would need before and after comparison to be swayed)